core/src/main/java/org/elasticsearch/action/support/replicatedbroadcast/TransportBroadcastReplicationAction.java

```diff
@@ -0,0 +1,145 @@
+
+    private final TransportReplicationAction replicatedBroadcastShardAction;
+    private final ClusterService clusterService;
+
+    public TransportBroadcastReplicationAction(String name, Class<Request> request, Settings settings, ThreadPool threadPool, ClusterService clusterService,
+                                               TransportService transportService,
+                                               ActionFilters actionFilters, IndexNameExpressionResolver indexNameExpressionResolver, TransportReplicationAction replicatedBroadcastShardAction) {
+        super(settings, name, threadPool, transportService, actionFilters, indexNameExpressionResolver,request);
+        this.replicatedBroadcastShardAction = replicatedBroadcastShardAction;
+        this.clusterService = clusterService;
+    }
+
+    @Override
+    protected void doExecute(final Request request, final ActionListener<Response> listener) {
+        GroupShardsIterator groupShardsIterator = clusterService.operationRouting().searchShards(clusterService.state(), indexNameExpressionResolver.concreteIndices(clusterService.state(), request), null, null);
+        final CopyOnWriteArrayList<ShardResponse> shardsResponses = new CopyOnWriteArrayList();
+        final CountDown responsesCountDown = new CountDown(groupShardsIterator.size());
+        if (responsesCountDown.isCountedDown() == false) {
+            for (final ShardsIterator shardsIterator : groupShardsIterator) {
+                final ShardRouting shardRouting = shardsIterator.nextOrNull();
+                if (shardRouting == null) {
+                    if (responsesCountDown.countDown()) {
+                        finishAndNotifyListener(listener, shardsResponses);
+                    }
+                } else {
+                    final ShardId shardId = shardRouting.shardId();
+                    replicatedBroadcastShardAction.execute(newShardRequest(request, shardId), new ActionListener<ShardResponse>() {
+                        @Override
+                        public void onResponse(ShardResponse shardResponse) {
+                            shardsResponses.add(shardResponse);
+                            if (responsesCountDown.countDown()) {
+                                logger.trace("replicated broadcast: got response from {}", shardId);
+                                finishAndNotifyListener(listener, shardsResponses);
+                            }
+                        }
+
+                        @Override
+                        public void onFailure(Throwable e) {
+                            logger.trace("replicated broadcast: got failure from {}", shardId);
+                            int totalNumCopies = clusterService.state().getMetaData().index(shardId.index().getName()).getNumberOfReplicas() + 1;
```

Comment:
can we cache the cluster state we used when we started this operation? this way we know for sure the index is still in it.